Implement FromStr

qmega opened the pull request:
My use case was for derive-based parsing with [bpaf](https://docs.rs/bpaf/latest/bpaf/).
Implementing FromStr allows usage like this:
```rs
#[derive(Clone, Bpaf)]
#[bpaf(options)]
struct Arguments {
    /// How long to wait before giving up (if not specified, wait forever)
    timeout: Option<DurationString>,
    ...
}
```
and working arg-parsing code gets generated automatically.

The existing TryFrom implementation had the same semantics so I just used that, but since that took ownership and FromStr only needs a reference (as does the actual conversion code), I made the old function defer to the new one.

Second commit is just fixing a typo (missing ")" in error message), making the 3 instances of the format pattern match (kept the one with all units in increasing length order) and putting it in a code block in the readme. If you don't like the code block, I'm not particularly attached to it.

RonniSkansing commented:
Hi, could I asks you to

- Add test for coverage of the new functionality.

Your are also welcome to 
- Bump the minor version and add yourself to the contributors list in Cargo

After pulling the changes, I will release the new version.

qmega commented:
What would you consider appropriate testing of the new interface, given the existing one calls the same code? Would you like me to change all the existing parse tests to try both interfaces and assert the desired result for each? I'd factor it out so each test would call something common with a string and expected Duration result. That's what makes most sense to me but want to confirm before changing a bunch of your existing tests.

I don't really think this contribution is significant enough to list me as an author, but if you'd like to put me there I won't object.

RonniSkansing commented:
Testing wise, I think that would be a great idea, this would make it easier if later on things divert. 

It is just a small lib and any contribution I will add as contributor unless asked otherwise :)

qmega commented:
Refactored tests as discussed. Threw an example of the FromStr-based usage into the readme and doctests too.

Bumped the version to 0.2.0 and added myself as an author. For the record, though, I don't consider this contribution original enough to be copyrightable; should you ever want to relicense this or something, you don't need to ask my permission if I haven't made any further contributions by then.

RonniSkansing commented:
Thanks. 

Merged and bumped on crates.io
